File: Crowmask.Library/Cache/IInteractionLookup.cs
namespace Crowmask.Library.Cache
{
    /// <summary>
    /// Provides a way to quickly look up a Crowmask post, given the ID of a
    /// relevant ActivityPub object (a Like, an Announce, or a reply). Allows
    /// for dependency inversion.
    /// </summary>
    public interface IInteractionLookup
    {
        /// <summary>
        /// Given the ID of a relevant ActivityPub object, yields the
        /// corresponding submission ID, or yields no items if the activity or
        /// object is not present on any submissions in Crowmask's cache.
        /// </summary>
        /// <param name="activity_or_reply_id">The ID of an Announce or Like activity, or the ID of a reply to an artwork submission</param>
        /// <returns>All relevant submission IDs (expected to be zero or one items)</returns>
        IAsyncEnumerable<int> GetRelevantSubmitIdsAsync(string activity_or_reply_id);

        /// <summary>
        /// Given the ID of a relevant ActivityPub object, yields the
        /// corresponding journal ID, or yields no items if the activity or
        /// object is not present on any journals in Crowmask's cache.
        /// </summary>
        /// <param name="activity_or_reply_id">The ID of an Announce or Like activity, or the ID of a reply to a journal</param>
        /// <returns>All relevant journal IDs (expected to be zero or one items)</returns>
        IAsyncEnumerable<int> GetRelevantJournalIdsAsync(string activity_or_reply_id);
    }
}

File: Crowmask.Interfaces/IInteractionLookup.vb
''' <summary>
''' Provides a way to quickly look up a Crowmask post, given the ID of a
''' relevant ActivityPub object (a Like, an Announce, or a reply).
''' </summary>
''' <param name="activity_or_reply_id"></param>
''' <returns></returns>
Public Interface IInteractionLookup
    ''' <summary>
    ''' Given the ID of an ActivityPub boost, like, or reply, returns the
    ''' corresponding submission IDs, or returns null if the activity or
    ''' object is not present on any submissions in Crowmask's cache. An
    ''' interaction will generally be tied to just one post.
    ''' </summary>
    ''' <param name="activity_or_reply_id">The ID of an Announce or Like activity, or the ID of a reply to an artwork submission</param>
    Function GetRelevantSubmitIdsAsync(activity_or_reply_id As String) As Task(Of IEnumerable(Of Integer))
End Interface
